File: client-sdk/src/services/governance.ts
/**
 * Governance Service - Multisig and Action Management
 * 
 * Provides governance controls for multi-signature approvals,
 * policy changes, and enforcement actions.
 */

import { AxiosInstance } from 'axios';
import { BaseService } from './base';
import { EventEmitter } from '../events';

// ═══════════════════════════════════════════════════════════════════════════════
// TYPES
// ═══════════════════════════════════════════════════════════════════════════════

export type ActionType = 
  | 'WALLET_PAUSE'
  | 'WALLET_UNPAUSE'
  | 'MANDATORY_ESCROW'
  | 'RELEASE_ESCROW'
  | 'POLICY_UPDATE'
  | 'WHITELIST_ADD'
  | 'WHITELIST_REMOVE'
  | 'BLACKLIST_ADD'
  | 'BLACKLIST_REMOVE'
  | 'EMERGENCY_STOP';

export type ActionStatus = 
  | 'PENDING'
  | 'AWAITING_SIGNATURES'
  | 'QUORUM_REACHED'
  | 'EXECUTED'
  | 'EXPIRED'
  | 'CANCELLED';

export type ActionScope = 
  | 'SINGLE_WALLET'
  | 'WALLET_CLUSTER'
  | 'GLOBAL';

export interface RiskContext {
  summary: string;
  fanOut?: number;
  velocityDeviation?: number;
  priorDisputes?: boolean;
  mlConfidence?: number;
  relatedCaseIds?: string[];
}

export interface Signature {
  signerId: string;
  signerRole: string;
  signerAddress: string;
  signature: string;
  signedAt: string;
  acknowledgedSnapshotHash: string;
  hashVerified: boolean;
  mfaMethod?: string;
  mfaVerifiedAt?: string;
}

export interface GovernanceAction {
  id: string;
  type: ActionType;
  status: ActionStatus;
  scope: ActionScope;
  targetAddress?: string;
  durationHours?: number;
  expiresAt: string;
  riskContext: RiskContext;
  isReversible: boolean;
  reversalConditions?: string;
  initiatedBy: string;
  initiatedAt: string;
  uiSnapshotHash: string;
  uiSnapshotId: string;
  policyVersion: string;
  requiredSignatures: number;
  signatures: Signature[];
}

export interface CreateActionRequest {
  type: ActionType;
  scope: ActionScope;
  targetAddress?: string;
  durationHours?: number;
  riskContext: RiskContext;
  uiSnapshotHash: string;
  policyVersion: string;
}

export interface SignActionRequest {
  actionId: string;
  signature: string;
  acknowledgedSnapshotHash: string;
  mfaToken?: string;
}

export interface SigningResult {
  success: boolean;
  actionId: string;
  signatureId?: string;
  currentSignatures: number;
  requiredSignatures: number;
  quorumReached: boolean;
  error?: string;
}

export interface ExecutionResult {
  success: boolean;
  actionId: string;
  transactionHash?: string;
  executedAt?: string;
  error?: string;
}

export interface WYASummary {
  actionType: string;
  scope: string;
  target: string;
  riskSummary: string;
  reversible: boolean;
  signatures: { current: number; required: number };
}

export interface ActionListOptions {
  status?: ActionStatus;
  type?: ActionType;
  limit?: number;
  offset?: number;
}

// ═══════════════════════════════════════════════════════════════════════════════
// SERVICE
// ═══════════════════════════════════════════════════════════════════════════════

export class GovernanceService extends BaseService {
  constructor(http: AxiosInstance, events: EventEmitter) {
    super(http, events);
  }

  /**
   * Create a new governance action
   */
  async createAction(request: CreateActionRequest): Promise<GovernanceAction> {
    const response = await this.http.post<GovernanceAction>('/governance/actions', request);
    this.events.emit('governance:action_created', response.data);
    return response.data;
  }

  /**
   * Get a governance action by ID
   */
  async getAction(actionId: string): Promise<GovernanceAction | null> {
    try {
      const response = await this.http.get<GovernanceAction>(`/governance/actions/${actionId}`);
      return response.data;
    } catch {
      return null;
    }
  }

  /**
   * List governance actions
   */
  async listActions(options?: ActionListOptions): Promise<GovernanceAction[]> {
    const response = await this.http.get<{ actions: GovernanceAction[] }>(
      '/governance/actions',
      { params: options }
    );
    return response.data.actions;
  }

  /**
   * Get pending actions for a user
   */
  async getPendingActions(userId: string): Promise<GovernanceAction[]> {
    const response = await this.http.get<GovernanceAction[]>(
      `/governance/actions/pending`,
      { params: { userId } }
    );
    return response.data;
  }

  /**
   * Sign a governance action
   */
  async signAction(request: SignActionRequest): Promise<SigningResult> {
    const response = await this.http.post<SigningResult>(
      `/governance/actions/${request.actionId}/sign`,
      request
    );
    
    this.events.emit('governance:signature_added', response.data);
    
    if (response.data.quorumReached) {
      this.events.emit('governance:quorum_reached', response.data);
    }
    
    return response.data;
  }

  /**
   * Execute a governance action (after quorum reached)
   */
  async executeAction(actionId: string): Promise<ExecutionResult> {
    const response = await this.http.post<ExecutionResult>(
      `/governance/actions/${actionId}/execute`
    );
    return response.data;
  }

  /**
   * Cancel a governance action
   */
  async cancelAction(actionId: string, reason: string): Promise<{ success: boolean }> {
    const response = await this.http.post(`/governance/actions/${actionId}/cancel`, { reason });
    return response.data;
  }

  /**
   * Get What-You-Approve summary for an action
   */
  async getWYASummary(actionId: string): Promise<WYASummary> {
    const response = await this.http.get<WYASummary>(`/governance/actions/${actionId}/wya`);
    return response.data;
  }

  /**
   * Get audit trail for an action
   */
  async getAuditTrail(actionId: string): Promise<Array<{
    event: string;
    timestamp: string;
    userId: string;
    details: Record<string, unknown>;
  }>> {
    const response = await this.http.get(`/governance/actions/${actionId}/audit`);
    return response.data.events;
  }

  /**
   * Check if user can sign an action
   */
  async canUserSign(actionId: string, userId: string): Promise<{
    canSign: boolean;
    reason?: string;
  }> {
    const response = await this.http.get(`/governance/actions/${actionId}/can-sign`, {
      params: { userId },
    });
    return response.data;
  }

  /**
   * Calculate quorum progress
   */
  static calculateQuorumProgress(action: GovernanceAction): {
    current: number;
    required: number;
    percentage: number;
  } {
    const current = action.signatures.length;
    const required = action.requiredSignatures;
    return {
      current,
      required,
      percentage: Math.min(100, (current / required) * 100),
    };
  }

  /**
   * Get action type label
   */
  static getActionTypeLabel(type: ActionType): string {
    const labels: Record<ActionType, string> = {
      WALLET_PAUSE: 'Pause Wallet',
      WALLET_UNPAUSE: 'Unpause Wallet',
      MANDATORY_ESCROW: 'Mandatory Escrow',
      RELEASE_ESCROW: 'Release Escrow',
      POLICY_UPDATE: 'Update Policy',
      WHITELIST_ADD: 'Add to Whitelist',
      WHITELIST_REMOVE: 'Remove from Whitelist',
      BLACKLIST_ADD: 'Add to Blacklist',
      BLACKLIST_REMOVE: 'Remove from Blacklist',
      EMERGENCY_STOP: 'Emergency Stop',
    };
    return labels[type] || type;
  }
}

File: client-sdk-python/amttp/services/governance.py
"""Governance Service — WYAS multi-sig action management."""

from __future__ import annotations

from typing import Any, Dict, List, Literal, Optional

from pydantic import BaseModel

from amttp.services.base import BaseService

ActionType = Literal["policy_override", "risk_threshold", "whitelist", "blacklist", "config_change"]
ActionStatus = Literal["pending", "approved", "rejected", "executed", "cancelled", "expired"]
ActionScope = Literal["global", "per_address", "per_policy"]


class RiskContext(BaseModel):
    risk_score: float = 0
    risk_level: str = "low"
    affected_address: Optional[str] = None


class Signature(BaseModel):
    signer: str = ""
    signature: str = ""
    signed_at: str = ""


class GovernanceAction(BaseModel):
    id: str = ""
    type: str = ""
    status: str = "pending"
    scope: str = "global"
    description: str = ""
    proposed_by: str = ""
    risk_context: Optional[RiskContext] = None
    signatures: List[Signature] = []
    required_signatures: int = 0
    parameters: Optional[Dict[str, Any]] = None
    created_at: str = ""
    executed_at: Optional[str] = None
    expires_at: Optional[str] = None


class CreateActionRequest(BaseModel):
    type: str
    scope: str = "global"
    description: str = ""
    proposed_by: str = ""
    parameters: Optional[Dict[str, Any]] = None
    required_signatures: int = 2


class SignActionRequest(BaseModel):
    signer: str
    signature: str


class SigningResult(BaseModel):
    action_id: str = ""
    signer: str = ""
    total_signatures: int = 0
    required_signatures: int = 0
    quorum_reached: bool = False


class ExecutionResult(BaseModel):
    action_id: str = ""
    executed: bool = False
    tx_hash: Optional[str] = None
    error: Optional[str] = None


class WYASSummary(BaseModel):
    total_actions: int = 0
    pending: int = 0
    approved: int = 0
    executed: int = 0
    rejected: int = 0
    avg_approval_time_hours: float = 0


# ── service ───────────────────────────────────────────────────────────────────


class GovernanceService(BaseService):
    """WYAS governance actions with multi-sig approval."""

    async def create_action(self, request: CreateActionRequest) -> GovernanceAction:
        """Propose a new governance action."""
        resp = await self._http.post(
            "/governance/actions", json=request.model_dump(exclude_none=True)
        )
        resp.raise_for_status()
        data = GovernanceAction.model_validate(resp.json())
        self._events.emit("governance:action_created", data.model_dump())
        return data

    async def get_action(self, action_id: str) -> GovernanceAction:
        """Get governance action details."""
        resp = await self._http.get(f"/governance/actions/{action_id}")
        resp.raise_for_status()
        return GovernanceAction.model_validate(resp.json())

    async def list_actions(
        self,
        *,
        status: Optional[str] = None,
        action_type: Optional[str] = None,
        limit: Optional[int] = None,
        offset: Optional[int] = None,
    ) -> List[GovernanceAction]:
        """List governance actions."""
        params: Dict[str, Any] = {}
        if status:
            params["status"] = status
        if action_type:
            params["type"] = action_type
        if limit is not None:
            params["limit"] = limit
        if offset is not None:
            params["offset"] = offset
        resp = await self._http.get("/governance/actions", params=params)
        resp.raise_for_status()
        return [GovernanceAction.model_validate(a) for a in resp.json().get("actions", [])]

    async def sign_action(self, action_id: str, request: SignActionRequest) -> SigningResult:
        """Sign a governance action."""
        resp = await self._http.post(
            f"/governance/actions/{action_id}/sign", json=request.model_dump()
        )
        resp.raise_for_status()
        data = SigningResult.model_validate(resp.json())
        self._events.emit("governance:signature_added", {
            "action_id": action_id, "signer": request.signer
        })
        if data.quorum_reached:
            self._events.emit("governance:quorum_reached", {"action_id": action_id})
        return data

    async def execute_action(self, action_id: str) -> ExecutionResult:
        """Execute an approved governance action."""
        resp = await self._http.post(f"/governance/actions/{action_id}/execute")
        resp.raise_for_status()
        data = ExecutionResult.model_validate(resp.json())
        if data.executed:
            self._events.emit("governance:action_executed", {"action_id": action_id})
        return data

    async def cancel_action(self, action_id: str) -> GovernanceAction:
        """Cancel a governance action."""
        resp = await self._http.post(f"/governance/actions/{action_id}/cancel")
        resp.raise_for_status()
        data = GovernanceAction.model_validate(resp.json())
        self._events.emit("governance:action_cancelled", {"action_id": action_id})
        return data

    async def get_summary(self) -> WYASSummary:
        """Get governance summary."""
        resp = await self._http.get("/governance/summary")
        resp.raise_for_status()
        return WYASSummary.model_validate(resp.json())

    async def health(self) -> Dict[str, Any]:
        """Check service health."""
        resp = await self._http.get("/governance/health")
        resp.raise_for_status()
        return resp.json()
